### ISSUE TITLE
Support for boxed primitive values?

### ISSUE THREAD

**Issue by ro0gr** [qunitjs/qunit#1596]
### TLDR;

should this succeed?

```js
assert.true(new Boolean(true));
```

----
I'm looking for a way to integrate page objects(briefly described in the #1570) with QUnit, without a need to deal with custom user assertions. As the end result I'd like to have a snippet like this working: 

```js
  assert.strictEqual(anInput.isFocused, true);
  assert.strictEqual(anInput.value, 'some value');
  assert.strictEqual(submitButton.isVisible, true);
```

In a case of failure, I expect to see some meaningful failure message(which is a point of the #1570). That's why I think expressing any page object property as a boxed primitive value is a promising direction to explore.

Under a boxed primitive, I mean an object wrapper with a possiblility to have some contextual info useful for building an assertion message, and the `valueOf(` returning an actual value from the DOM. Here is a simplified example of what I mean:

```js
const myInput = buildAPageObjectInstance({
  selector: '...',
  
  isFocused: {
    valueOf() {
      return this.domElement === document.activeElement;
    }
  },
  
  value: {
    valueOf() {
      return this.domElement.value;
    }
  }
})
```

For now this kind of page objects do only work(mostly) for the `assert.equal(`, due to the `==`+`valueOf(`, and fails with `strictEqual` for the obvious reason:

```js
'string value' !== { 
  valueOf() { 
    return 'string value'; 
  }
};
```

A solution could be just to use an `assert.equal(`, but this is not perfect because of the values coercion, like `true == 1`.

If ignore my page objects case, we could think of another use case for a boxed primitive values:

```js
assert.true(new Boolean(true));
```

Right now, this leads to a failure. Though, at this point, it's not obviously clear to me if this is a correct behavior. In my opinion, it may have sense to succeed, cause I assert against the `Boolean` equal to `true`. If such, this idea could be applied to the rest of assertions, so the `strictEqual(` will support boxed primitve values, like: 

```js
const boxedString = { 
  valueOf() { return 'some string';}
}
assert.stringEqual(boxedString, 'some string');
```

I totally understand this quite a controversial suggestion, and I can't see such a behavior for neither of NodeJS asserts, chai, jest. However, it doesn't necessarily mean this is a bad thing to do in general. 

As far as I can see, taking into account boxed primitive values should not have any breaking impact on existing test suites, and opens a door for the new external integrations opportunities.

Would appreciate any thoughts on the topic. 

Thanks!

**Comment by Krinkle** [comment]
I'll discose my bias and say that I think boxed primitives are an anti-pattern. We do support them to the same extent that JavaScript supports them natively, such as in loose equality operations. As such, the non-strict equality methods will work, e.g. `assert.equal( obj, 'expected' );`.

If these classes were a part of the application under test, I would recommend treating them as methods on classes like any other class, e.g. `assert.strictEqual( obj.toString(), 'expected' );`. After all, it seems bad practice to leave ambiguity over whether something is a primitive or a class like this. The explicit call seems desirable even.

If I understand correctly from https://github.com/qunitjs/qunit/issues/1570, I think these are source code classes then, but helper classes that exist specifically for use within the test suite, and that you'd like to formulate some extra information with a failing test based on information from these helper classes. I think that's an important use case and I want to make sure that this is address properly.

But, I think the approach of using `valueOf()` is not going to work out well for either of us in the long-term. Suppose we do change strictEqual to be slightly less strict, and define some detectable subset of objects that we believe are unlikely to be matched by accident, and let the assertion methods call `.valueOf()` on those first. That by itself would not adress https://github.com/qunitjs/qunit/issues/1570, as one still has to make use of the object somewhere by making other (currently unsupported) changes to QUnit internals.

If you are fine with override internals, however, then one could also overwrite the assertion methods to make `assert.strictEqual`, `assert.true`, etc. special case instances of `MyHelperClass` and use `.valueOf()` for the comparison. Of course these override would need to live somewhere, but I think that might be okay as they could presumably live in the same place as the helper classes and the other would-be override.

Alternatively (per https://github.com/qunitjs/qunit/issues/1570) one could define them under a different name if preferred, e.g. `assert.myStrictEqual` and use that for strings, numbers, and booleans alike; or define `assert.myTrue/False` along with it.

Happy to continue at https://github.com/qunitjs/qunit/issues/1570, and I actually have at least one more idea I'll share there.

I'll close this ticket for now as I don't consider it a general improvement for QUnit, but also believe it would not address your higher-level problem by itself, and that the other changes needed on your end already touch areas that present alternative solutions within the supported API.
